api/jobs_test.go

```diff
@@ -316,3 +341,80 @@ func ExpectedJob(id string,
+
+		jobStoreMock := &apiMock.JobStorerMock{
+			PutNumberOfTasksFunc: func(ctx context.Context, id string, count int) error {
+				switch id {
+				case validJobID2:
+					return nil
+				default:
+					return errors.New("the job store does not contain the job id entered")
+				}
+			},
+			AcquireJobLockFunc: func(ctx context.Context, id string) (string, error) {
+				switch id {
+				case unLockableJobID:
+					return "", errors.New("acquiring lock failed")
+				default:
+					return "", nil
+				}
+			},
+		}
+
+		apiInstance := api.Setup(ctx, mux.NewRouter(), jobStoreMock)
+
+		Convey("When a request is made to update the number of tasks of a specific job that exists in the Job Store", func() {
+			req := httptest.NewRequest("PUT", fmt.Sprintf("http://localhost:25700/jobs/%s/number_of_tasks/%s", validJobID2, validCount), nil)
+			resp := httptest.NewRecorder()
+
+			apiInstance.Router.ServeHTTP(resp, req)
+
+			Convey("Then a status code 200 is returned", func() {
+				So(resp.Code, ShouldEqual, http.StatusOK)
+			})
+		})
+
+		Convey("When a request is made to update the number of tasks of a specific job that does not exist in the Job Store", func() {
+			req := httptest.NewRequest("PUT", fmt.Sprintf("http://localhost:25700/jobs/%s/number_of_tasks/%s", notFoundJobID, validCount), nil)
+			resp := httptest.NewRecorder()
+
+			apiInstance.Router.ServeHTTP(resp, req)
+
+			Convey("Then job resource was not found returning a status code of 404", func() {
+				So(resp.Code, ShouldEqual, http.StatusNotFound)
+				errMsg := strings.TrimSpace(resp.Body.String())
+				So(errMsg, ShouldEqual, "Failed to find job in job store")
+			})
+		})
+
+		Convey("When a request is made to update the number of tasks but the path parameter given as the Count is not an integer", func() {
+			req := httptest.NewRequest("PUT", fmt.Sprintf("http://localhost:25700/jobs/%s/number_of_tasks/%s", validJobID2, invalidCount), nil)
+			resp := httptest.NewRecorder()
+
+			apiInstance.Router.ServeHTTP(resp, req)
+
+			Convey("Then it is a bad request returning a status code of 400", func() {
+				So(resp.Code, ShouldEqual, http.StatusBadRequest)
+				errMsg := strings.TrimSpace(resp.Body.String())
+				So(errMsg, ShouldEqual, "invalid path parameter - count should be an integer")
+			})
+		})
+
+		Convey("When a request is made to update the number of tasks but the Job Store is unable to lock the id", func() {
+			req := httptest.NewRequest("PUT", fmt.Sprintf("http://localhost:25700/jobs/%s/number_of_tasks/%s", unLockableJobID, validCount), nil)
+			resp := httptest.NewRecorder()
+
+			apiInstance.Router.ServeHTTP(resp, req)
+
+			Convey("Then an error with status code 500 is returned", func() {
+				So(resp.Code, ShouldEqual, http.StatusInternalServerError)
+				errMsg := strings.TrimSpace(resp.Body.String())
+				So(errMsg, ShouldEqual, expectedServerErrorMsg)
+			})
+		})
```

Comment:
Question: what happens if count is negative? Granted its unlikely but that value is provided to the service so its wise to validate its OK rather than assume its OK. 

If setting a negative value would be considered incorrect or invalid can we add another test case to prove this is handled correctly?

Reply:
good idea - done